nupic/research/frameworks/vernon/experiments/cl_experiment.py

```diff
@@ -141,6 +156,162 @@ def compute_task_indices(cls, config, dataset):
                 task_indices[i].extend(class_indices[j + (i * num_classes_per_task)])
         return task_indices
 
+    def get_active_classes(self, task_num):
+        """
+        Returns a list of label indices that are "active" during training under the
+        continual learning scenario specified by the config. In the "task" scenario,
+        only the classes that are being trained are active. In the "class" scenario,
+        all tasks that the model has previously observed are active. More information
+        about active classes can be found here: https://arxiv.org/abs/1904.07734
+
+        :param task_num: zero-based index of the task
+        """
+        if self.cl_experiment_type == "task":
+            return [label for label in range(
+                self.num_classes_per_task * task_num,
+                self.num_classes_per_task * (task_num + 1)
+            )]
+        elif self.cl_experiment_type == "class":
+            return [label for label in range(
+                self.num_classes_per_task * (task_num + 1)
+            )]
+        elif self.cl_experiment_type == "domain":
+            raise NotImplementedError
+
+    def train_cl_model(
```

Comment:
I'm curious why this function and `evaluate_cl_model` is needed. I believe they may not be. Ideally, we should avoid duplicate code whenever possible. Is it possible to call `set_active_tasks` on the samplers during `run_task`? This would look something like
```
    def run_task(self):
    
        # Configure the sampler to load only samples from classes of the active task.
        active_classes = self.get_active_classes(self.current_task)
        self.train_loader.sampler.set_active_tasks(active_classes)
        self.val_loader.sampler.set_active_tasks(active_classes)

        # Run epochs
        for _ in range(self.epochs):
            self.run_epoch()
```

Reply:
It's already using the sampler to set active tasks, but the task and class scenarios treat output nodes separately when evaluating the loss.

E.g., in splitMNIST, the task scenario computes the output for all 10 units, but only the outputs of task 0 (the output heads for classes 0 & 1) are updated during the backward pass.

Reply:
Just saw this, apologies for the repeated comment below. 

If it is a task scenario, the output will have two units and the target will be one hot encoded to reflect that, so they will match. If it is a class scenario, you only want the distribution to be between the classes which actually show in the dataset. So why not backpropagate from the other nodes? They will never show as target, so we will push them to zero, which is what is supposed to happen in that case. 

Can you please point to a theoretical reference so I can understand this a little bit better.

Reply:
It seems the definition comes from this paper: https://arxiv.org/pdf/1904.07734v1.pdf. I will give my understanding below, please correct me if I am wrong, but it might be better if we talk over zoom.

In Task-IL, it is a multiheaded scenario. You are informed of which task it is before hand and you can switch heads, to use a head dedicated to that task. In this PR I don't see where the heads are switched.

In Class-IL, it is a single head. You are not informed of the task beforehand. If you mask the output, that means you have the task information (which is not supposed to be available). My understanding is that in Class-IL, given the information of which task we are in is not available, we have to backpropagate from all nodes.

Reply:
A bit more info from the appendix of that paper. 

For Task-IL: " normalization performed by the final softmax layer only takes into account these active nodes, and that learning is thus restricted to those nodes". Masking output when calculating gradients is not enough, softmax will still be computed over all nodes in the forward pass.  

For Class-IL: "With the Class-IL scenario only classes from up to the previous task can be predicted.". This makes sense, and it means the previous implementation is incorrect, since the only active nodes are the ones from classes seen so far (before we were calculating for all nodes). It is also aligned with the code in `get_active_classes` in this PR. However, the softmax calculation is also incorrect in this scenario.

Reply:
An easier way to implement this is in Task-IL, keeping track of the heads, and appending the proper head to the network when task is switched. No need to change train or eval. 

In Class-IL, start with the number of heads for the present task, and incrementally add new nodes to the single head as more tasks are presented. As before, can add the new nodes when the task is switched, no need to change train or eval.

Reply:
I'm  basing my implementation off the one in the paper that you linked. Here is how they select output units to compute the loss, just as we do: [link](https://github.com/GMvandeVen/continual-learning/blob/master/encoder.py#L185). Also, since that relies on a variable `active_classes`, here is where they compute it, just like us: [link](https://github.com/GMvandeVen/continual-learning/blob/master/train.py#L77-L82)

Reply:
There is no issues on how you filter y or compute active classes - the only issue in this implementation on the PR is how softmax is calculated. 

Another option you can consider is taking softmax out of the network, and filtering the logits before calculating the softmax. That can avoid having to "switch" the heads of the network, but it might require rewriting the train and eval functions, so it will reuse less code.